hindsight/io_modules/derived_stream.lua

```diff
@@ -0,0 +1,197 @@
+        local buffer_size   = read_config("buffer_size") or 10000 * 1024
+        assert(buffer_size > 0, "buffer_size must be greater than zero")
+
+        local last_insert   = 0
+        local uuid          = nil
+
+        local env = assert(driver.postgres())
+        local con = assert(env:connect(db_config.name, db_config.user, db_config._password, db_config.host, db_config.port))
+
+        local function get_output_file(ts)
+            local day = math.floor(ts / (SEC_IN_DAY * 1e9))
+            local f = files[day]
+            if not f then
+                local date = os.date("%Y%m%d", ts / 1e9)
+                local table_name = string.format("%s_%s", name, date)
+                local filename = string.format("%s/%s_%s.sql", buffer_path, cfg_name, date)
+                local ok, cnt, err = pcall(con.execute, con, rs.get_create_table_sql(table_name, schema))
+                if ok and err and not string.match(err, "duplicate key violates unique constraint") then
+                    error(err) -- non recoverable database error
+                end
+                if not ok then return nil, err end -- API error
+
+                f = {fh = nil, table_name = table_name, filename = filename, offset = 0}
+                files[day] = f
+            end
+
+            if not f.fh then
+                f.fh = assert(io.open(f.filename, "w+"))
+                f.offset = 0
+            end
+            return f
+        end
+
+        local function insert_files() -- all files are flushed at once to ensure the checkpoint is accurate
+            for k,v in pairs(files) do
+                if v.fh and v.offset ~= 0 then
+                    v.fh:seek("set")
+                    local ok, cnt, err = pcall(con.execute, con, v.fh:read("*a")) -- read the entire file and execute the query
+                    if ok and err then -- database error
+                        error(err)
+                    end
+                    if not ok then return err end -- API error
+                    v.fh:close()
+                    v.fh = nil
+                    os.remove(v.filename);
+                end
+            end
+            last_insert = os.time()
+        end
+
+        local function process_message()
+            local file
+            if not (uuid and uuid == read_message("Uuid")) then -- make sure we aren't in a retry loop
+                file, err = get_output_file(read_message(ts_field))
+                if not file then return -3, err end
+
+                uuid = nil
+                if file.offset == 0 then
+                    file.fh:write("INSERT INTO ", file.table_name, " VALUES ")
+                else
+                    file.fh:write(",")
+                end
+                rs.write_values_sql(file.fh, con, schema)
+                file.offset = file.fh:seek("end")
+            end
+
+            if not file or file.offset >= buffer_size then
```

Comment:
Yup, verified that the `file.fh:seek("end")` on L102 returns `nil` in error cases, which falls through to cause the compare on L105 to fail.